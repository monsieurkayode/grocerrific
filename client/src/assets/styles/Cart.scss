.modal {
  .modal__main {
    .checkout-items {
      max-height: px-to-rem(380px);
      overflow-y: scroll;
    }

    .cart-item-container {
      border-bottom: 1px solid $light-grey-1;
      padding: 20px 0;
    }

    .error {
      color: tomato;
    }

    .success {
      color: teal;
    }

    .checkout-status {
      font-size: 12px;
      padding: 4px 0;
      font-family: $font-family-prompt;
    }

    .checkout-item-wrapper {
      @extend %flex__container_row;
      @extend %nowrap;
      justify-content: space-between;
      align-items: center;
    }

    .total {
      border-bottom: none;
      padding: 20px 0;
    }

    .item-name, #total {
      font-family: $font-family-roboto;
      font-weight: 500;
      color: $light-grey-6;
    }

    .item-name {
      width: 45%;
      font-size: 14.5px;
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      text-transform: capitalize;
    }

    #total {
      text-transform: uppercase;
      font-size: 20px;
      font-weight: 700;
    }

    .order {
      width: 25%;
    }

    .price {
      flex: 0 1 25%;
      margin-right: 20px;
      text-align: right;
      text-overflow: ellipsis;
      overflow: hidden;
    }

    .price, .total-price {
      color: #4d75a1;
      font-family: $font-family-roboto;
      font-weight: 500;
    }

    .total-price {
      font-size: 20px;
    }

    .inc, .dec {
      padding: 5px 10px;
      font-size: 16px;
      cursor: pointer;
      outline: none;
      border: none;
      font-weight: 900;
      background: transparent;

      &:disabled {
        cursor: not-allowed;
        color: $light-grey-4;
      }
    }

    .quantity {
      width: 40px;
      margin: 0 5px;
      border-radius: 5px;
      outline: none;
      padding: 4px 8px;
      border: 1px solid $light-grey-1;

      &:focus {
        border: 1px solid #4d75a1;
      }
    }

    .remove {
      cursor: pointer;
      outline: none;
    }

    .buy {
      float: right;
      border: 2px solid #4d75a1;
      background: transparent;
      padding: px-to-rem(10px) px-to-rem(50px);
      color: #4d75a1;
      font-size: px-to-rem(14px);
      font-family: $font-family-prompt;
      border-radius: px-to-rem(50px);
      text-shadow: 1px 2px 8px rgba(78, 45, 126, 0.3);
      transition: all 0.8s cubic-bezier(0.075, 0.82, 0.165, 1);
      cursor: pointer;
      outline: none;
      user-select: none;

      &:disabled {
        cursor: not-allowed;
        border: 2px solid #4d75a1;
        background: #4d75a1;
        color: #FFF;
      }

      &:hover:not(:disabled) {
        border: 2px solid #4d75a1;
        background: #4d75a1;
        color: #FFF;
      }
    }
  }
}
