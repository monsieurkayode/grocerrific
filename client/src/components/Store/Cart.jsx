import React from 'react';
import {
  func,
  shape,
  arrayOf,
  bool
} from 'prop-types';

import Modal from '../common/Modal';
import CartItem from './CartItem';

const Cart = ({
  closeModal,
  cartItems,
  removeFromCart,
  checkout,
  checkingOut
}) => {
  const computeTotal = () => (
    cartItems.reduce(
      (acc, cartItem) => acc + (cartItem.price * cartItem.quantity), 0
    )
  );

  return (
    <Modal
      closeModal={closeModal}
      title="grocery cart"
    >
      <div className="checkout-items">
        {cartItems.map(cartItem => (
          <CartItem
            key={cartItem.id}
            removeFromCart={removeFromCart}
            {...cartItem}
            cartItems={cartItems}
            checkingOut={checkingOut}
          />
        ))}
      </div>
      <div className="checkout-item-wrapper total">
        <p id="total">Total</p>
        <span className="total-price">&#8358;{computeTotal()}</span>
      </div>
      <div className="checkout__button">
        <button
          className="buy uppercase"
          type="button"
          onClick={() => checkout(cartItems)}
          disabled={checkingOut}
        >
          {checkingOut ? 'Processing...' : 'Checkout'}
        </button>
      </div>
    </Modal>
  );
};


Cart.propTypes = {
  closeModal: func.isRequired,
  removeFromCart: func.isRequired,
  checkout: func.isRequired,
  checkingOut: bool.isRequired,
  cartItems: arrayOf(shape({})).isRequired
};

export default Cart;
