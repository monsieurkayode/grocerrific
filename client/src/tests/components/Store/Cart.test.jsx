import React from 'react';
import { shallow } from 'enzyme';
import Cart from '../../../components/Store/Cart';
import mockCartItems from '../../__mocks__/mockCartItems';

const setup = () => {
  const props = {
    cartItems: mockCartItems,
    closeModal: jest.fn(),
    removeFromCart: jest.fn(),
    checkout: jest.fn(() => Promise.resolve()),
    checkingOut: false,
  };

  const wrapper = shallow(<Cart {...props} />);

  return { wrapper, props };
};

describe('Cart Component', () => {
  const { wrapper, props } = setup();

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render checkout button', () => {
    expect(wrapper.find('.buy').length).toBe(1);
  });

  it('checkout button should have text Processing... while checkingOut', () => {
    wrapper.setProps({ checkingOut: true });
    expect(wrapper.find('.buy').text()).toEqual('Processing...');
  });

  it('should call checkout when the checkout button is clicked', () => {
    const checkoutButton = wrapper.find('.buy');
    checkoutButton.simulate('click');
    expect(props.checkout).toHaveBeenCalled();
  });
});
