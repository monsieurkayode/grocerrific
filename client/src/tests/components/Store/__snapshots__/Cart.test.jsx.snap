// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component renders correctly 1`] = `
<Modal
  closeModal={[MockFunction]}
  title="grocery cart"
>
  <div
    className="checkout-items"
  >
    <Connect(CartItem)
      cartItems={
        Array [
          Object {
            "id": "5bf4754223602e00211d6587",
            "maxQuantity": 4,
            "name": "mamador",
            "price": 1250,
            "quantity": 3,
          },
          Object {
            "id": "5bf4754223602e00211d6588",
            "maxQuantity": 15,
            "name": "sausage",
            "price": 120,
            "quantity": 12,
          },
          Object {
            "id": "5bf4754223602e00211d6589",
            "maxQuantity": 10,
            "name": "salad creme",
            "price": 1500,
            "quantity": 1,
          },
        ]
      }
      checkingOut={false}
      id="5bf4754223602e00211d6587"
      key="5bf4754223602e00211d6587"
      maxQuantity={4}
      name="mamador"
      price={1250}
      quantity={3}
      removeFromCart={[MockFunction]}
    />
    <Connect(CartItem)
      cartItems={
        Array [
          Object {
            "id": "5bf4754223602e00211d6587",
            "maxQuantity": 4,
            "name": "mamador",
            "price": 1250,
            "quantity": 3,
          },
          Object {
            "id": "5bf4754223602e00211d6588",
            "maxQuantity": 15,
            "name": "sausage",
            "price": 120,
            "quantity": 12,
          },
          Object {
            "id": "5bf4754223602e00211d6589",
            "maxQuantity": 10,
            "name": "salad creme",
            "price": 1500,
            "quantity": 1,
          },
        ]
      }
      checkingOut={false}
      id="5bf4754223602e00211d6588"
      key="5bf4754223602e00211d6588"
      maxQuantity={15}
      name="sausage"
      price={120}
      quantity={12}
      removeFromCart={[MockFunction]}
    />
    <Connect(CartItem)
      cartItems={
        Array [
          Object {
            "id": "5bf4754223602e00211d6587",
            "maxQuantity": 4,
            "name": "mamador",
            "price": 1250,
            "quantity": 3,
          },
          Object {
            "id": "5bf4754223602e00211d6588",
            "maxQuantity": 15,
            "name": "sausage",
            "price": 120,
            "quantity": 12,
          },
          Object {
            "id": "5bf4754223602e00211d6589",
            "maxQuantity": 10,
            "name": "salad creme",
            "price": 1500,
            "quantity": 1,
          },
        ]
      }
      checkingOut={false}
      id="5bf4754223602e00211d6589"
      key="5bf4754223602e00211d6589"
      maxQuantity={10}
      name="salad creme"
      price={1500}
      quantity={1}
      removeFromCart={[MockFunction]}
    />
  </div>
  <div
    className="checkout-item-wrapper total"
  >
    <p
      id="total"
    >
      Total
    </p>
    <span
      className="total-price"
    >
      â‚¦
      6690
    </span>
  </div>
  <div
    className="checkout__button"
  >
    <button
      className="buy uppercase"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Checkout
    </button>
  </div>
</Modal>
`;
